iterators/1-each.rb

```diff
@@ -0,0 +1,144 @@
+rainbow = %w[red orange yellow green gray indogo violet]
+
+# use cicle for
+for color in rainbow
+  puts color
+end
+
+# red
+# orange
+# yellow
+# green
+# gray
+# indogo
+# violet
+
+puts
+# use each
+rainbow.each { |color| puts color }
+
+# times
+5.times {|t| puts t}
+
+# 0
+# 1
+# 2
+# 3
+# 4
+
+puts
+# upto and downto
+5.upto(10) {|t| puts t}
+
+15.downto(10) {|t| puts t}
+
+# Iterator            Synonym
+# each
+# each_with_index
+# map                 collect
+# select              find_all
+# reject              delete_if
+# reduce              inject
+# each_with_object
+
+puts
+# each, no mutation
+arr = [1, 2, 3, 4, 5]
+
+res = arr.each { |x| puts x ** 2 }
+p res
+1
+4
+9
+16
+25
+[1, 2, 3, 4, 5] # returns arr
+
+puts
+# each_with_index
+rainbow = %w[red orange yellow green gray indogo violet]
+rainbow.each_with_index { |color, ind| puts "#{ind}: #{color}" }
+
+# 0: red
+# 1: orange
+# 2: yellow
+# 3: green
+# 4: gray
+# 5: indogo
+# 6: violet
+
+puts 
+# map, return new collection
+res = [1, 2, 3, 4, 5].map { |x| x +1 }
+p res 
```

Comment:
**Note** that the value that the block returns is used to form the elements of the new array. 

- [ ] Therefore, if you leave the puts method in the block, the elements will be displayed on separate lines, but the resulting the array will contain only nil values - the result returned by the puts method 

```
result = [1, 2, 3, 4, 5].map { |x| puts x + 1 }
p result # [nil, nil, nil, nil, nil]
```

**The map** iterator has a synonym `collect` . 

- [ ] The collect form is less common in programs, map is shorter and therefore more popular.

```
p [1, 2, 3, 4, 5].collect { |x| x + 1 } # [2, 3, 4, 5, 6]
```
[see more: ](https://www.rubyguides.com/2018/10/ruby-map-method/#:~:text=Ruby%20Map%20vs%20Collect,most%20common%20version%20is%20map%20.)